core/temporal/src/fields.rs

```diff
@@ -463,6 +480,144 @@ impl TemporalFields {
 
         Ok(())
     }
+
+    /// Merges two `TemporalFields` values given a specific `CalendarSlot`.
+    pub fn merge_fields<C: CalendarProtocol>(
+        &self,
+        other: &Self,
+        calendar: &CalendarSlot<C>,
+    ) -> TemporalResult<Self> {
+        let add_keys = other.keys().collect::<Vec<_>>();
+        let overridden_keys = calendar.field_keys_to_ignore(&add_keys)?;
+
+        let mut result = Self::default();
+
+        for key in self.keys() {
+            let value = if overridden_keys.contains(&key) {
+                other.get(&key)
+            } else {
+                self.get(&key)
+            };
+
+            if let Some(value) = value {
+                result.set_field_value(&key, &value)?;
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+/// Iterator over `TemporalFields` keys.
+pub struct KeyIter {
```

Comment:
Just a thought: Maybe we should rename this to just `Keys` in the [standard library they use this convention](https://doc.rust-lang.org/std/collections/btree_map/struct.Keys.html?search=Keys). Same for `ValuesIter`.